addons/stock_landed_costs/models/stock_landed_costs.py

```diff
@@ -0,0 +1,273 @@
+                              quantity=qty_out,
+                              account_id=already_out_account_id)
+            credit_line = dict(credit_line,
+                               name=(line.name + ": " + str(qty_out) + _(' already out')),
+                               quantity=qty_out,
+                               account_id=debit_account_id)
+            diff = diff * qty_out / line.quantity
+            if diff > 0:
+                debit_line['debit'] = diff
+                credit_line['credit'] = diff
+            else:
+                # negative cost, reverse the entry
+                debit_line['credit'] = -diff
+                credit_line['debit'] = -diff
+            AccountMoveLine.create(debit_line)
+            AccountMoveLine.create(credit_line)
+        move.assert_balanced()
+        return True
+
+    def _create_account_move(self):
+        return self.env['account.move'].create({
+            'journal_id': self.account_journal_id.id,
+            'date': self.date,
+            'ref': self.name
+        })
+
+    def _check_sum(self):
+        """
+        Will check if each cost line its valuation lines sum to the correct amount
+        and if the overall total amount is correct also
+        """
+        correctcost = {}
+        total = 0
+        for valuation_line in self.valuation_adjustment_line_ids:
+            if correctcost.get(valuation_line.cost_line_id):
+                correctcost[valuation_line.cost_line_id] += valuation_line.additional_landed_cost
+            else:
+                correctcost[valuation_line.cost_line_id] = valuation_line.additional_landed_cost
+            total += valuation_line.additional_landed_cost
+
+        prec = self.env['decimal.precision'].precision_get('Account')
+        # float_compare returns 0 for equal amounts
+        res = not bool(float_compare(total, self.amount_total, precision_digits=prec))
+        for costl in correctcost.keys():
+            if float_compare(correctcost[costl], costl.price_unit, precision_digits=prec):
+                res = False
+        return res
+
+    @api.multi
+    def button_validate(self):
+        for cost in self:
+            if cost.state != 'draft':
+                raise UserError(_('Only draft landed costs can be validated'))
+            if not cost.valuation_adjustment_line_ids or not cost._check_sum():
+                raise UserError(_('You cannot validate a landed cost which has no valid valuation adjustments lines. Did you click on Compute?'))
+            move = cost._create_account_move()
+
+            for line in cost.valuation_adjustment_line_ids.filtered(lambda line: line.move_id):
+                qty_out = 0
+                per_unit = line.final_cost / line.quantity
+                diff = per_unit - line.former_cost_per_unit
+                for quant in line.move_id.quant_ids:
+                    quant.cost += diff
```

Comment:
Be aware of using Active Record Pattern, as each assignment will trigger write on database .